src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,168 @@
+
+    private static class Node<T> {
+        T item;
+        Node<T> next;
+        Node<T> prev;
+
+        Node(T item, Node<T> next, Node<T> prev) {
+            this.item = item;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public MyLinkedList() {
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        Node<T> l = tail;
+        Node<T> node = new Node<T>(value, null, l);
+        tail = node;
+        if (l == null) {
+            head = node;
+        } else {
+            l.next = node;
+        }
+        ++size;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index);
+        if (index == size) {
+            linkToEnd(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        return node(index).item;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> node = node(index);
+        node.item = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        return unlink(node(index));
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        Node<T> node;
+        T value = null;
+        if (t == null) {
+            for (node = head; node != null; node = node.next) {
+                if (node.item == null) {
+                    value = unlink(node);
+                }
+            }
+        } else {
+            for (node = head; node != null; node = node.next) {
+                if (t.equals(node.item)) {
+                    value = unlink(node);
+                }
+            }
+        }
+        return value;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void linkToEnd(T t) {
+        Node<T> l = tail;
+        Node<T> node = new Node<T>(t, null, l);
+        tail = node;
+        if (l == null) {
+            head = node;
+        } else {
+            l.next = node;
+        }
+        ++size;
+    }
+
+    private Node<T> node(int index) {
```

Comment:
```suggestion
    private Node<T> getNode(int index) {
```

we can optimize this method if we add checking:
if(size /2 < index){
...
} else {
...
}